extensions/ql-vscode/src/remote-queries/view/FileCodeSnippet.tsx

```diff
@@ -0,0 +1,224 @@
+      return 'blue';
+    case 'Warning':
+      return warningColor;
+    case 'Error':
+      return 'red';
+  }
+};
+
+const replaceSpaceChar = (text: string) => text.replaceAll(' ', '\u00a0');
+
+const shouldHighlightLine = (lineNumber: number, highlightedRegion: HighlightedRegion) => {
+  if (lineNumber < highlightedRegion.startLine) {
+    return false;
+  }
+
+  if (highlightedRegion.endLine == undefined) {
+    return lineNumber == highlightedRegion.startLine;
+  }
+
+  return lineNumber <= highlightedRegion.endLine;
+};
+
+const Container = styled.div`
+  font-family: ui-monospace, SFMono-Regular, SF Mono, Menlo, Consolas, Liberation Mono, monospace;
+  font-size: x-small;
+  width: 55em;
+`;
+
+const TitleContainer = styled.div`
+  border: 0.1em solid ${borderColor};
+  border-top-left-radius: 0.2em;
+  border-top-right-radius: 0.2em;
+  padding: 0.5em;
+`;
+
+const CodeContainer = styled.div`
+  border-left: 0.1em solid ${borderColor};
+  border-right: 0.1em solid ${borderColor};
+  border-bottom: 0.1em solid ${borderColor};
+  border-bottom-left-radius: 0.2em;
+  border-bottom-right-radius: 0.2em;
+  padding-top: 1em;
+  padding-bottom: 1em;
+`;
+
+const MessageText = styled.div`
+  font-size: x-small;
+  padding-left: 0.5em;
+`;
+
+const MessageContainer = styled.div`
+  padding-top: 0.5em;
+  padding-bottom: 0.5em;
+`;
+
+const PlainLine = ({ text }: { text: string }) => {
+  return <span>{replaceSpaceChar(text)}</span>;
+};
+
+const HighlightedLine = ({ text }: { text: string }) => {
+  return <span style={{ backgroundColor: highlightColor }}>{replaceSpaceChar(text)}</span>;
+};
+
+const Message = ({
+  messageText,
+  currentLineNumber,
+  highlightedRegion,
+  borderColor,
+  children
+}: {
+  messageText: string,
+  currentLineNumber: number,
+  highlightedRegion: HighlightedRegion,
+  borderColor: string,
+  children: React.ReactNode
+}) => {
+  if (highlightedRegion.startLine !== currentLineNumber) {
+    return <></>;
+  }
+
+  return <MessageContainer>
+    <Box
+      borderColor="border.default"
+      borderWidth={1}
+      borderStyle="solid"
+      borderLeftColor={borderColor}
+      borderLeftWidth={3}
+      paddingTop="1em"
+      paddingBottom="1em">
+      <MessageText>
+        {messageText}
+        {children && <>
+          <VerticalSpace size={2} />
+          {children}
+        </>
+        }
+      </MessageText>
+    </Box>
+
+  </MessageContainer>;
+};
+
+const CodeLine = ({
+  line,
+  lineNumber,
+  highlightedRegion
+}: {
+  line: string,
+  lineNumber: number,
+  highlightedRegion: HighlightedRegion
+}) => {
+  if (!shouldHighlightLine(lineNumber, highlightedRegion)) {
+    return <PlainLine text={line} />;
+  }
+
+  const isSingleLineHighlight = highlightedRegion.endLine === undefined;
+  const isFirstHighlightedLine = lineNumber === highlightedRegion.startLine;
+  const isLastHighlightedLine = lineNumber === highlightedRegion.endLine;
+
+  const highlightStartColumn = isSingleLineHighlight
```

Comment:
Going back to our conversation on the previous PR about handling invalid source locs, there are a number of common errors that I think you can handle right here:

```js
if (highlightStartColumn > highilightEndColumn) {
  ([highlightStartColumn, highilightEndColumn] = [highilightEndColumn, highlightStartColumn]);
}

highlightStartColumn = Math.max(highlightStartColumn, 0);
highilightEndColumn = Math.max(highlightStartColumn, highilightEndColumn);
```

Or some variant of this. It looks like this code won't cause an error if the locations are invalid, but they might cause missing highlights. Again, this is pretty rare, but it does happen.

Reply:
Thanks for the example/code. Lets discuss validation next week!